pkg/restore/restore.go

```diff
@@ -696,7 +748,40 @@ func (ctx *context) restoreResource(resource, namespace, resourcePath string) (a
 
 			// Check if the PV exists in the cluster before attempting to create
 			// a volume from the snapshot, in order to avoid orphaned volumes (GH #609)
-			_, err := resourceClient.Get(name, metav1.GetOptions{})
+			pv, err := ctx.pvClient.Get(name, metav1.GetOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				addToResult(&errs, namespace, fmt.Errorf("error checking existence for PV %s: %v", name, err))
+			}
+
+			if isPVReleasedOrBound(pv) {
+				ctx.log.Infof("Persistent volume %s still in use, waiting for deletion", name)
+				timeout := time.After(ctx.resourceTerminatingTimeout)
+				ticker := time.NewTicker(time.Second)
+				defer ticker.Stop()
+			loop:
+				for {
```

Comment:
@skriss PTAL. I'm not feeling super great about the for/select combo, but maybe it's my unfamiliarity with the Go channels mechanisms.

2 things I'm noticing with this and the namespace loop (L513):

1. Need some way to honor ctrl-c or other graceful shutdown. Should we hook it in to [`signals.CancelOnShutdown`](https://github.com/heptio/ark/blob/master/pkg/cmd/server/server.go#L269) somehow?
2. Do we need to find some way to mark the restore as `Failed` for the NS/PV timeout cases? Or is completed with errors ok? For reference, this is what a restore that hits the timeout looks like currently:

```
x1c in /home/nrb/go/src/github.com/heptio/ark (git) fix-691 U
% ark restore describe voltest-20181109132929
Name:         voltest-20181109132929
Namespace:    heptio-ark
Labels:       <none>
Annotations:  <none>

Backup:  voltest

Namespaces:
  Included:  *
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io, backups.ark.heptio.com, restores.ark.heptio.com
  Cluster-scoped:  auto

Namespace mappings:  <none>

Label selector:  <none>

Restore PVs:  auto

Phase:  Completed

Validation errors:  <none>

Warnings:
  Ark:        <none>
  Cluster:    <none>
  Namespaces:
    nginx-example:  not restored: persistentvolumeclaims "nginx-logs" already exists and is different from backed up version.
                    not restored: pods "nginx-deployment-f9d69f8fd-c6k6g" already exists and is different from backed up version.
                    not restored: deployments.apps "nginx-deployment" already exists and is different from backed up version.
                    not restored: endpoints "my-nginx" already exists and is different from backed up version.
                    not restored: replicasets.apps "nginx-deployment-f9d69f8fd" already exists and is different from backed up version.
                    not restored: services "my-nginx" already exists and is different from backed up version.

Errors:
  Ark:      Timed out waiting on PV pvc-98f7668e-e390-11e8-9753-42010a96003b to be deleted
  Cluster:    <none>
  Namespaces: <none>
```

Reply:
For #2, completed with errors is what we'd expect.

Looking at #1.

Reply:
For #1, I'm not sure that we actually want to do this - would like to think some more and maybe @ncdc will find time to provide input.  We don't currently have any code that will stop a backup or restore in the middle if a shutdown signal is received - we try to finish the backup/restore, then the controllers/server exit.  We'd want to be able to properly mark the restore's status as failed/interrupted, which would take more code that we don't currently have.

If we were going to implement it, looks like the most straight-forward way to do this is in `server.go`, around L590, pass `ctx` as a new arg to the `restore.NewKubernetesRestorer` func.  Have that func set it as a field on the `kubernetesRestorer`.  In the `Restore` method on the `kubernetesRestorer`, pass that `ctx` down to the `restoreCtx` that's created (probably as a field).  Then, in your `select` statements that are doing the polling, add an additional case that receives on `ctx.Done()` and exits -- though we'd probably want to end up marking restores that were exited early like this as Failed.

Reply:
Thanks. That makes sense. My use case was lowering the timeout for testing (default of 10, but wanted to drop to 1 or 2), so doing a `kill -9` on it I guess isn't too terrible.

I can leave that part as is and address the remaining comments and update tests.